apps/ejabberd/src/mod_last_riak.erl

```diff
@@ -0,0 +1,76 @@
+%%%----------------------------------------------------------------------
+%%% File    : mod_last.erl
+%%% Author  : Pawel Pikula <pawel.pikula@erlang-solutions.com>
+%%% Purpose : mod_last mnesia backend (XEP-0012)
+%%%
+%%%
+%%% ejabberd, Copyright (C) 2002-2014   ProcessOne
+%%% MongooseIM, Copyright (C) 2014      Erlang Solutions Ltd.
+%%%
+%%%----------------------------------------------------------------------
+
+-module(mod_last_riak).
+
+-behaviour(mod_last).
+
+-include("mod_last.hrl").
+-include("ejabberd.hrl").
+
+-define(TIMESTAMP_IDX, {integer_index, "timestamp"}).
+
+%% API
+-export([init/2,
+         get_last/2,
+         count_active_users/2,
+         set_last_info/4,
+         remove_user/2]).
+
+-spec init(ejabberd:server(), list()) ->ok.
+init(_VHost, _Opts) ->
+    %% we are using commong riak pool
+    ok.
+
+-spec get_last(ejabberd:luser(), ejabberd:lserver()) ->
+    {ok, non_neg_integer(), binary()} | {error, term()} | not_found.
+get_last(LUser, LServer) ->
+    case mongoose_riak:get(bucket_type(LServer), LUser) of
+        {ok, Obj} ->
+            Status = riakc_obj:get_value(Obj),
+            MD = riakc_obj:get_update_metadata(Obj),
+            [Timestamp] = riakc_obj:get_secondary_index(MD, ?TIMESTAMP_IDX),
+            {ok, Timestamp, Status};
+        {error, notfound} ->
+            not_found;
+        {error, Reason} ->
+            {error, Reason}
+    end.
+
+-spec count_active_users(ejabberd:lserver(), non_neg_integer()) -> non_neg_integer().
+count_active_users(LServer, TimeStamp) ->
+    Idx = {index, bucket_type(LServer), ?TIMESTAMP_IDX, TimeStamp+1, infinity()},
+    RedMF = {modfun, riak_kv_mapreduce, reduce_count_inputs},
+    Red = [{reduce, RedMF, [{reduce_phase_batch_size, 1000}], true}],
+    {ok, [{0, [Count]}]}  = mongoose_riak:mapred(Idx, Red),
```

Comment:
Yes, it is given as example [here](http://docs.basho.com/riak/latest/dev/using/2i/#Querying) section  **Count Bucket Objects via $bucket Index**. 

I think the best way to find which way is better is to test it with at least several millions of entries. After discussion with @fenek we suspect, that the only difference will be who will consume more resources, whether will it be MIM  or Riak.

In case of mapred we receive just count - aggregation and actual counting is done on Riak's side. The *get_index_range* function gives us list of keys(which may be very big) and MiM has to calculate the length. It is not that bad, because Riak allows to paginate these results, so we can avoid receiving 10 millions keys at once, by doing many many requests. I guess the mapred approach might be more efficient, but this is my guess only...